frontend/components/tree/js/tree.service.ts

```diff
@@ -1027,51 +1034,70 @@ export class TreeService {
 		this.ViewerService.clearHighlights();
 		this.DocsService.closeDocs();
 
-		if (this.currentSelectedNodes.length) {
-			while (this.currentSelectedNodes.length) {
-				const currentNode = this.currentSelectedNodes.pop();
-				currentNode.selected = this.SELECTION_STATES.unselected;
-				this.setSelectionOnParentNodes(currentNode, this.SELECTION_STATES.unselected);
+		const visitedParents = [];
+		const visitedChildren = [];
+
+		while (this.currentSelectedNodes.length) {
+			const currentNode = this.currentSelectedNodes.pop();
+			currentNode.selected = this.SELECTION_STATES.unselected;
+
+			// Skip over any node that has already been visited
+			if (visitedChildren.indexOf(currentNode) === -1) {
+
+				// Skip over any parent that has been visited
+				if (visitedParents.indexOf(currentNode) === -1) {
+					const parents = this.setSelectionOnParentNodes(currentNode, this.SELECTION_STATES.unselected);
```

Comment:
You are still visiting some parents multiple times here. Let's use the following example:

currentlySelectedNodes = ["A", "B"];
A's treePath is C D E F A
B's treePath is C D E F A B

loop#1 dealing with A:
`setSelectionOnParentNodes` would be processing C D E F
loop#2 dealing with B:
`setSelectionOnParentNodes` would be processing C D E F A (because visitedParents does not have B in it, C D E F A are in seen parents, but we don't check for it.)

I would just give up on using `setSelectionOnParentNodes` here, you have to loop through the parents to push them into visited parents anyway, you may as well just set their visibility here. less looping, less traversal, and you can break once you've found an ancestor you have visited.